### PR TITLE
Input digest tagger

### PR DESCRIPTION
**Pull request by piotr-szybicki**
Implementation of inputDigest tagger from the [Improve taggers proposal](https://github.com/GoogleContainerTools/skaffold/blob/master/docs/design_proposals/digest-tagger.md) This PR is much bigger than I originally intended. But whichever way I looked I was hitting the circular dependencies. So the first 2 commits are about extracting `build.ArtifactGraph` and `build.Artifact` into its own package. 

It does mostly what internal skaffold hasher is doing a minus couple of things:

* Doesn't hash the build arguments (I don't believe they are not relevant. )
* Doesn't hash test dependencies (I don't believe they are not relevant. )

Some areas to think of (maybe in this or in future PR's):
* following target in multi-stage docker builds
* including image cotainerId in the tag. (But to get there the implementation of docker.GetDependencies would have to change including its contract of the method)
* documentation what exactly is being used in the tag calculation (how only COPY and ADD commands can influence tag)

### CODE REVIEW

**Comment by briandealwis** [comment]
Sorry for the delay in reviewing this @piotr-szybicki.  I'll try to look today — this particular issue has been on my wishlist for a bit.

**Review by briandealwis — COMMENTED** [review]
Hi @piotr-szybicki — I must apologize profusely for taking so long to review this PR.  It generally looks great and is exactly what I was hoping for.

The one big quibble is the name of `pkg/skaffold/build/dep`.  The team talked about this and we'd like to move this package up in the hierarchy and call it `graph` (so `pkg/skaffold/graph`).

If you're sick of this PR, I'm happy to take this on.

**Review by piotr-szybicki — COMMENTED** [review]
> Hi @piotr-szybicki — I must apologize profusely for taking so long to review this PR. It generally looks great and is exactly what I was hoping for.
> 
> The one big quibble is the name of `pkg/skaffold/build/dep`. The team talked about this and we'd like to move this package up in the hierarchy and call it `graph` (so `pkg/skaffold/graph`).
> 
> If you're sick of this PR, I'm happy to take this on.

Hi, no problem has been busy with other stuff myself and didn't chase. I looked over the comments and they make total sense I will implement all the changes you requested. I will also add some documentation as generally understanding how this works is not that trivial as I found out. It's different for docker comparing to Bazel or jib.

File: cmd/skaffold/app/cmd/run.go
Comment (piotr-szybicki) [reply cmd/skaffold/app/cmd/run.go:47]: How that got there is beyond me.

File: pkg/skaffold/build/tag/input_digest.go
Comment (piotr-szybicki) [reply pkg/skaffold/build/tag/input_digest.go]: yea, too much copy-paste without thinking is never a good thing.

File: pkg/skaffold/build/tag/input_digest.go
Comment (piotr-szybicki) [reply pkg/skaffold/build/tag/input_digest.go]: absolutely. Do you think I should also remove it from the hash.fileHasher()? Feels like the internal scaffold cache and the tagger should be consistent.

File: pkg/skaffold/build/tag/input_digest.go
Comment (piotr-szybicki) [reply pkg/skaffold/build/tag/input_digest.go]: SO there is no `filepath.ToPath()` function I used `filepath.Clean()` that seems to do the job. Also, Do you think I should change it in the `hash.fileHasher()`? For consistency.

File: pkg/skaffold/build/tag/input_digest_test.go
Comment (piotr-szybicki) [reply pkg/skaffold/build/tag/input_digest_test.go]: Done

File: pkg/skaffold/build/tag/input_digest.go
Comment (piotr-szybicki) [reply pkg/skaffold/build/tag/input_digest.go]: Ehh, never mind, they will never be consistent. I just took a look at the `hash.go` code and see that you are also hashing the build arguments. Totally make sense if you consider the development life cycle. Something that I don't think is appropriate to be a part of the image hash (Might be wrong on that one). I mean the way I see it we want to be close to what docker is doing with it's `layer.DiffID`.

File: pkg/skaffold/build/tag/input_digest_test.go
Comment (piotr-szybicki) [reply pkg/skaffold/build/tag/input_digest_test.go]: done

**Review by briandealwis — COMMENTED** [review]
Thanks @piotr-szybicki!  I've put in a few comments.  If we can get these addressed, then it'll be ready for commit.

#5492 (and #5504) brought to light that we need to incorporate more than just the inputs — and this input tagger should use the [artifact-hashing mechanism](https://github.com/GoogleContainerTools/skaffold/blob/master/pkg/skaffold/build/cache/hash.go#L59) that's used elsewhere in Skaffold.  Unfortunately this artifact-hasher is not yet exposed in an easy-to-consume manner ([mechanism](https://github.com/GoogleContainerTools/skaffold/blob/master/pkg/skaffold/build/cache/lookup.go#L36), [lister](https://github.com/GoogleContainerTools/skaffold/blob/master/pkg/skaffold/runner/new.go#L82-L86)).  This is a bigger change and can be done separately if you don't want to take it on (just let me know).

File: pkg/skaffold/build/tag/tag.go
Comment (nkubala) [reply pkg/skaffold/build/tag/tag.go]: done

File: cmd/skaffold/app/cmd/run.go
Comment (nkubala) [reply cmd/skaffold/app/cmd/run.go]: done

File: pkg/skaffold/build/tag/input_digest.go
Comment (nkubala) [reply pkg/skaffold/build/tag/input_digest.go]: done

File: pkg/skaffold/build/tag/input_digest.go
Comment (nkubala) [reply pkg/skaffold/build/tag/input_digest.go]: done

File: pkg/skaffold/build/tag/input_digest.go
Comment (nkubala) [reply pkg/skaffold/build/tag/input_digest.go]: done

File: pkg/skaffold/build/tag/input_digest.go
Comment (nkubala) [reply pkg/skaffold/build/tag/input_digest.go]: done

File: pkg/skaffold/build/tag/tag.go
Comment (nkubala) [reply pkg/skaffold/build/tag/tag.go]: done, though this one was a little strange because `GenerateTag` is entangled with a few functions that already use `*latest.Artifact`. did the best I could and I think it's a little cleaner now

File: pkg/skaffold/build/tag/input_digest.go
Comment (nkubala) [reply pkg/skaffold/build/tag/input_digest.go]: I have very limited understanding of how this encoder works, but the `json.Encoder` that's being used takes the entire set of files and returns a single encoding based on that set. is it possible to encode each file individually and then combine the encodings to achieve the same result?

it's worth noting that the elements of `inputs` are the *hashes* of each file and its contents, not the entire file contents themselves. so while there could be thousands of files, this would be a set of thousands of hashes, each of which is only a few bytes.

**Comment by nkubala** [comment]
@piotr-szybicki I took the liberty of rebasing your PR (I made some changes that made the merge conflicts quite annoying for you) - I hope this is ok!

I also addressed a few of @briandealwis's last comments - mostly formatting and cleaning suggestions, though I made one change to the arguments to `GenerateTag`. please let me know if all of these changes are ok, and if so (pending @briandealwis's final review) we'll merge this!

**Comment by piotr-szybicki** [comment]
@nkubala this is absolutely fine. Thanks for doing it. I was about to do this myself but my full time job is killing me lately.

**Comment by piotr-szybicki** [comment]
> Thanks @piotr-szybicki! I've put in a few comments. If we can get these addressed, then it'll be ready for commit.
> 
> #5492 (and #5504) brought to light that we need to incorporate more than just the inputs — and this input tagger should use the [artifact-hashing mechanism](https://github.com/GoogleContainerTools/skaffold/blob/master/pkg/skaffold/build/cache/hash.go#L59) that's used elsewhere in Skaffold. Unfortunately this artifact-hasher is not yet exposed in an easy-to-consume manner ([mechanism](https://github.com/GoogleContainerTools/skaffold/blob/master/pkg/skaffold/build/cache/lookup.go#L36), [lister](https://github.com/GoogleContainerTools/skaffold/blob/master/pkg/skaffold/runner/new.go#L82-L86)). This is a bigger change and can be done separately if you don't want to take it on (just let me know).

@briandealwis Sorry for the late response. But a year of my work is coming to a culmination release over coming weekend. I actually knew about this issue If you take a look at the original comment I made when creating this PR 
`including image cotainerId in the tag. (But to get there the implementation of docker.GetDependencies would have to change including its contract of the method)`
I have a fix on a branch so I will create a different PR to address this particular issue. Because this one got big enough as it is.

File: docs/content/en/schemas/v2beta13.json
Comment (briandealwis) [inline docs/content/en/schemas/v2beta13.json:2848]: This doesn't seem right

File: pkg/skaffold/build/tag/input_digest.go
Comment (piotr-szybicki) [reply pkg/skaffold/build/tag/input_digest.go]: the input array contains only a file paths, I iterate over them in the encode method hashing as you said content and relative path. Maybe the name is misleading.

File: pkg/skaffold/kubernetes/owner_test.go
Comment (briandealwis) [inline pkg/skaffold/kubernetes/owner_test.go:62]: The dep -> graph changes in this file seem a bit odd, but I don't think it really matters.

File: pkg/skaffold/schema/latest/config.go
Comment (briandealwis) [inline pkg/skaffold/schema/latest/config.go:182]: This should have type `InputDigest`!

File: pkg/skaffold/tag/tag.go
Comment (briandealwis) [inline pkg/skaffold/tag/tag.go:47]: Could inline this (but feel free to ignore)
```suggestion
		tag, err := v.GenerateTag(latest.Artifact{ImageName: DeprecatedImageName})
```

File: pkg/skaffold/tag/input_digest_test.go
Comment (briandealwis) [inline pkg/skaffold/tag/input_digest_test.go:34]: I guess this should test our implementation!   Though if we fail to detect a change, this won't help! 😆

**Review by briandealwis — APPROVED** [review]
LGTM once the InputDigest type is fixed up

File: pkg/skaffold/tag/tag.go
Comment (piotr-szybicki) [reply pkg/skaffold/tag/tag.go]: changed.

File: pkg/skaffold/schema/latest/config.go
Comment (piotr-szybicki) [reply pkg/skaffold/schema/latest/config.go]: yea, what a dumb mistake. Fixed.

File: pkg/skaffold/kubernetes/owner_test.go
Comment (piotr-szybicki) [reply pkg/skaffold/kubernetes/owner_test.go]: it's golang getting aggressive with it's renaming. Fixed.

File: docs/content/en/schemas/v2beta13.json
Comment (piotr-szybicki) [reply docs/content/en/schemas/v2beta13.json:2848]: fixed, by changing to correct type.

**Comment by piotr-szybicki** [comment]
I fix all the comments, and tests. I of the Travis CI Jobs has flaked out on me. Can anyone rerun?

**Comment by tejal29** [comment]
@piotr-szybicki sorry for the delay. do you have some cycles to rebase your PR. I would love to merge this.

**Comment by nkubala** [comment]
@tejal29 i'm on it :)

**Comment by nkubala** [comment]
all merge conflicts resolved and CI finally showing all green, so I'm going to merge this! @piotr-szybicki thanks so much for the contribution and all of your patience!

**Comment by mikekamornikov** [comment]
@nkubala As I see `InputDigest` is not supported by `CustomTemplateTagger` (see `CreateComponents` in `pkg/skaffold/tag/tagger_mux.go`):
```yaml
apiVersion: skaffold/v2beta14
kind: Config
profiles:
- name: test-custom-template
  build:
    tagPolicy:
      customTemplate:
        template: "release-aaa-{{.DIGEST}}"
        components:
        - name: DIGEST
          inputDigest: {}
    artifacts:
    - image: quay.io/my/project
      context: ../../..
      bazel:
        target: //project:docker.tar
``` 

```sh
$ skaffold build -p test-custom-template
creating runner: creating tagger: creating tagger: creating components: unknown component for custom template: DIGEST {GitTagger:<nil> ShaTagger:<nil> EnvTemplateTagger:<nil> DateTimeTagger:<nil> CustomTemplateTagger:<nil> InputDigest:0x7a479b8}
. If above error is unexpected, please open an issue https://github.com/GoogleContainerTools/skaffold/issues/new to report this error.
```

Is it by design? Should I file a bug?

**Comment by piotr-szybicki** [comment]
I just reread `improve digest proposal`. And yes I clearly missed `{{.DIGEST}}` when implementing. So filing a bug would be a good choice. I will find some time during the weekend to fix this.

**Pull request merged**
